### PR TITLE
Engagement trend view and table displays engagement activities.

### PR DESCRIPTION
**Pull request by dsjen**
- Refactored enrollment-trend-view to trend-view that displays multiple trends.
- Using activities call on the api-client instead of deprecated recent_activity call.
- Forum activities for trends behind feature gate.

The engagement activity trend chart tooltip will be updated with AN-3362.

engagement-1
engagement-2

@rocha @clintonb

### CODE REVIEW

**Comment by clintonb** [comment]
Looks like you forgot to mock some API calls. Tests are failing.

File: acceptance_tests/__init__.py [hunk 1]
```diff
@@ -47,6 +47,28 @@ def test_footer(self):
...
+
+    def assertTableColumnHeadingsEqual(self, table_selector, headings):
+        rows = self.page.q(css=('%s thead th' % table_selector))
+        self.assertTrue(rows.present)
+        self.assertListEqual(rows.text, headings)
+
+    def assertElementHasContent(self, css):
+        element = self.page.q(css=css)
+        self.assertTrue(element.present)
+        graph_html = element.html[0]
```
Comment (clintonb) [inline acceptance_tests/__init__.py]: Minor: Use a more generic variable name (e.g. `html`).

File: analytics_dashboard/courses/presenters.py [hunk 2]
```diff
@@ -39,49 +42,58 @@ def format_date(self, date):
...
-            course_id (str): ID of the course to retrieve summary information of.
+        Retrieve engagement activity trends for specified date range and return
+        results with zeros filled in for all activities.
         """
-        course = self.client.courses(course_id)
+        api_trends = self.client.courses(self.course_id).activity(start_date=start_date, end_date=end_date)
 
-        any_activity = course.recent_activity(AT.ANY)
+        # feature gate posted_forum.  If enabled, the forum will be rendered in the engagement page
+        trend_types = ['any', 'played_video', 'attempted_problem']
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py]: Use activity types from API client.

File: analytics_dashboard/courses/presenters.py [hunk 3]
```diff
@@ -39,49 +42,58 @@ def format_date(self, date):
...
+            trend_types.append('posted_forum')
 
-        # store our activity data from the API
-        activities = [any_activity, ]
+        # fill in gaps in activity with zero for display
+        trends = []
+        for datum in api_trends:
+            trend_week = {'weekEnding': self.format_date(self.parse_date_time_as_date(datum['interval_end']))}
+            for trend_type in trend_types:
+                trend_week[trend_type] = datum[trend_type] if trend_type in datum else 0
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py:95]: Why is this necessary? The data from the pipeline should already have this fix (if necessary).

File: analytics_dashboard/courses/presenters.py [hunk 4]
```diff
@@ -39,49 +42,58 @@ def format_date(self, date):
...
-                activity = course.recent_activity(activity_type)
-            except NotFoundError:
-                # We know the course exists, since we have gotten this far in the code, but
-                # there is no data for the specified activity type. Report it as null.
-                activity = {'activity_type': activity_type, 'count': None}
+    def get_summary(self):
+        """
+        Retrieve all summary numbers and week ending time.
+        """
+        course = self.client.courses(self.course_id)
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py]: Set this in the constructor and utilize `self.course` in `get_summary` and `get_trend_data`.

File: analytics_dashboard/courses/presenters.py [hunk 5]
```diff
@@ -39,49 +42,58 @@ def format_date(self, date):
...
 
-            activities.append(activity)
+        # store our activity data from the API
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py:104]: This comment doesn't reflect the actions of the code immediately below it.

File: analytics_dashboard/courses/templates/courses/engagement_content.html [hunk 6]
```diff
@@ -32,6 +27,25 @@ <h4 class="section-title">
         </span>
       </div>
     </div>
+
+    <div class="section-content section-data-graph">
+      <div class="section-content section-data-viz">
+        <div id="engagement-trend-view">
+            <div class="line-chart-container"><div class="line-chart ">
+                {% comment %}Translators: "Trend" is a graph of weekly student engagement. Please translate accordingly.{% endcomment %}
```
Comment (clintonb) [inline analytics_dashboard/courses/templates/courses/engagement_content.html]: Student activity. Engagement has multiple meanings in English.

File: analytics_dashboard/courses/templates/courses/engagement_content.html [hunk 7]
```diff
@@ -32,6 +27,25 @@ <h4 class="section-title">
...
+                {% trans "Loading Trend..." as loading_msg %}
+                {% include "loading.html" with message=loading_msg %}
+            </div></div>
+        </div>
+      </div>
+    </div>
+
+  <h4 class="section-title">
+    <span class="section-title-value">Student Activity</span>
```
Comment (clintonb) [inline analytics_dashboard/courses/templates/courses/engagement_content.html]: Mark for translation.

File: analytics_dashboard/courses/templates/courses/engagement_content.html [hunk 8]
```diff
@@ -32,6 +27,25 @@ <h4 class="section-title">
...
+    <span class="section-title-note small">What are my students interacting with?</span>
```
Comment (clintonb) [inline analytics_dashboard/courses/templates/courses/engagement_content.html]: Mark for translation.

File: analytics_dashboard/courses/tests/test_presenters.py [hunk 9]
```diff
@@ -1,46 +1,58 @@
...
-        raise NotFoundError
+def mock_activity_data():
+    # AT.POSTED_FORUM deliberately left off to test edge case where activity is empty
+    activity_types = [AT.ANY, AT.ATTEMPTED_PROBLEM, AT.PLAYED_VIDEO]
 
-    activity_types = [AT.ANY, AT.ATTEMPTED_PROBLEM, AT.PLAYED_VIDEO, AT.POSTED_FORUM]
-
-    summaries = {}
+    summaries = {
+        'interval_end': '2014-01-01T123456',
```
Comment (clintonb) [inline analytics_dashboard/courses/tests/test_presenters.py]: Minor: Our intervals are always at midnight (`000000`).

File: analytics_dashboard/courses/tests/test_presenters.py [hunk 10]
```diff
@@ -49,18 +61,36 @@ def test_get_summary(self):
...
-        self.assertIsNone(summary[AT.POSTED_FORUM], None)
+        self.assertIsNone(summary[AT.POSTED_FORUM])
+
+    @mock.patch('analyticsclient.course.Course.activity', mock.Mock(side_effect=mock_activity_trend_data))
+    def test_get_trend_data(self):
+        Switch.objects.create(name='show_engagement_forum_activity', active=True)
+        trends = self.presenter.get_trend_data()
+
+        # check to see if we get the right values back and that None is filled to 0
+        self.assertEqual(trends[0]['weekEnding'], '2014-01-01')
```
Comment (clintonb) [inline analytics_dashboard/courses/tests/test_presenters.py]: This would be a lot cleaner with `self.assertDictEquals()`.

File: analytics_dashboard/courses/tests/test_views.py [hunk 11]
```diff
@@ -182,36 +214,24 @@ def test_engagement_page_success(self):
         self.assertEqual(response.context['summary'][AT.PLAYED_VIDEO], 1000)
         self.assertEqual(response.context['summary'][AT.POSTED_FORUM], 0)
 
+        # check to make sure the activity trends are correct
+        trends = response.context['js_data']['course']['engagementTrends']
+        self.assertEqual(len(trends), 2)
+        self.assertEqual(trends[0]['weekEnding'], '2013-1-8')
```
Comment (clintonb) [inline analytics_dashboard/courses/tests/test_views.py]: Use `assertDictEqual`.

File: analytics_dashboard/courses/tests/test_views.py [hunk 12]
```diff
@@ -367,6 +387,27 @@ def test_response_no_data(self, mock_enrollment):
         self.assertEqual(response.content, data)
 
 
+class CourseEngagementActivityTrendCSVViewTests(CourseEngagementViewTestMixin, TestCase):
```
Comment (clintonb) [inline analytics_dashboard/courses/tests/test_views.py]: Given that we have three test cases that all test CSV views, it's time to refactor the code into a mixin. The base class has the body of `test_response`. Each subclass can override and mock where necessary.

File: analytics_dashboard/courses/tests/utils.py [hunk 13]
```diff
@@ -71,9 +71,28 @@ def set_empty_permissions(user):
...
         AT.ATTEMPTED_PROBLEM: 301,
         AT.PLAYED_VIDEO: 1000,
         AT.POSTED_FORUM: 0,
     }
+
+
+def mock_engagement_activity_trend_data():
+    return [
+        {
+            'weekEnding': '2013-1-8',
```
Comment (clintonb) [inline analytics_dashboard/courses/tests/utils.py]: Missing zeroes.

File: analytics_dashboard/courses/views.py [hunk 14]
```diff
@@ -283,9 +283,7 @@ def get_context_data(self, **kwargs):
 
         try:
             summary = presenter.get_summary()
-            end_date = summary['date']
-            start_date = end_date - datetime.timedelta(days=60)
-            end_date = end_date + datetime.timedelta(days=1)
+            start_date, end_date = get_date_range(summary['date'], 60)
             data = presenter.get_trend_data(start_date=start_date, end_date=end_date)
```
Comment (clintonb) [inline analytics_dashboard/courses/views.py]: This can be cleaned up. If no `end_date` passed, `get_trend_data` can assume a 60 day interval and do the math.

File: analytics_dashboard/courses/views.py [hunk 15]
```diff
@@ -283,9 +283,7 @@ def get_context_data(self, **kwargs):
 
         try:
             summary = presenter.get_summary()
-            end_date = summary['date']
-            start_date = end_date - datetime.timedelta(days=60)
-            end_date = end_date + datetime.timedelta(days=1)
+            start_date, end_date = get_date_range(summary['date'], 60)
```
Comment (clintonb) [inline analytics_dashboard/courses/views.py]: `60` should be stored in a constant that can be easily modified if we decide to change the display range in the future.

File: analytics_dashboard/courses/views.py [hunk 16]
```diff
@@ -345,10 +343,16 @@ def get_context_data(self, **kwargs):
...
         }
 
-        presenter = CourseEngagementPresenter()
-        summary = presenter.get_summary(self.course_id)
+        presenter = CourseEngagementPresenter(self.course_id)
+        try:
+            summary = presenter.get_summary()
+            start_date, end_date = get_date_range(summary['interval_end'], 60)
+            trends = presenter.get_trend_data(start_date=start_date, end_date=end_date)
+        except NotFoundError:
```
Comment (clintonb) [inline analytics_dashboard/courses/views.py]: This is not necessary. `CourseView` will handle this.

File: analytics_dashboard/courses/views.py [hunk 17]
```diff
@@ -283,9 +283,7 @@ def get_context_data(self, **kwargs):
 
         try:
             summary = presenter.get_summary()
-            end_date = summary['date']
-            start_date = end_date - datetime.timedelta(days=60)
-            end_date = end_date + datetime.timedelta(days=1)
+            start_date, end_date = get_date_range(summary['date'], 60)
             data = presenter.get_trend_data(start_date=start_date, end_date=end_date)
         except NotFoundError:
             raise Http404
```
Comment (clintonb) [inline analytics_dashboard/courses/views.py]: This is not necessary. `CourseView` will handle this.

File: analytics_dashboard/courses/views.py [hunk 18]
```diff
@@ -345,10 +343,16 @@ def get_context_data(self, **kwargs):
...
+        try:
+            summary = presenter.get_summary()
+            start_date, end_date = get_date_range(summary['interval_end'], 60)
+            trends = presenter.get_trend_data(start_date=start_date, end_date=end_date)
+        except NotFoundError:
+            raise Http404
 
-        summary['week_of_activity'] = get_formatted_date_time(summary['interval_end'])
+        context['js_data']['course']['engagementTrends'] = trends
+        summary['week_of_activity'] = (end_date - datetime.timedelta(days=1)).strftime('%B %d, %Y')
```
Comment (clintonb) [inline analytics_dashboard/courses/views.py]: Django is already wired for date formatting via `settings.DATE_FORMAT`. Instead of formatting here, use the [`date` filter](https://docs.djangoproject.com/en/dev/ref/templates/builtins/#date).

File: acceptance_tests/__init__.py (hunk 1 above)
Comment (dsjen) [inline acceptance_tests/__init__.py]: Make sense.  Done!

File: analytics_dashboard/courses/presenters.py (hunk 2 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py]: Done!

File: analytics_dashboard/courses/presenters.py (hunk 3 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py:95]: I asked about the case that there is no data for a particular activity, but there is data for other activities for a given week, and I thought the result is that the field won't be returned.  Is that not the case?  If so, I do love deleting code!

Comment (clintonb) [inline analytics_dashboard/courses/presenters.py:95]: Ah. Yes, that is correct.

File: analytics_dashboard/courses/presenters.py (hunk 4 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py]: Ok, I've set course as a member variable and use it instead.

File: analytics_dashboard/courses/presenters.py (hunk 3 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py:95]: I wanted to delete some code. :(

File: analytics_dashboard/courses/templates/courses/engagement_content.html (hunk 6 above)
Comment (dsjen) [inline analytics_dashboard/courses/templates/courses/engagement_content.html]: Ok, updated.

File: analytics_dashboard/courses/templates/courses/engagement_content.html (hunk 7 above)
Comment (dsjen) [inline analytics_dashboard/courses/templates/courses/engagement_content.html]: Done!

File: analytics_dashboard/courses/templates/courses/engagement_content.html (hunk 8 above)
Comment (dsjen) [inline analytics_dashboard/courses/templates/courses/engagement_content.html]: Done!

File: analytics_dashboard/courses/tests/test_presenters.py (hunk 9 above)
Comment (dsjen) [inline analytics_dashboard/courses/tests/test_presenters.py]: ok, update.

File: analytics_dashboard/courses/tests/test_presenters.py (hunk 10 above)
Comment (dsjen) [inline analytics_dashboard/courses/tests/test_presenters.py]: Good call.

File: analytics_dashboard/courses/tests/test_views.py (hunk 11 above)
Comment (dsjen) [inline analytics_dashboard/courses/tests/test_views.py]: Done!

File: analytics_dashboard/courses/tests/utils.py (hunk 13 above)
Comment (dsjen) [inline analytics_dashboard/courses/tests/utils.py]: Ok, add them in.

File: analytics_dashboard/courses/tests/test_views.py (hunk 12 above)
Comment (dsjen) [inline analytics_dashboard/courses/tests/test_views.py]: Done!

File: analytics_dashboard/courses/views.py (hunk 16 above)
Comment (dsjen) [inline analytics_dashboard/courses/views.py]: Removed.

File: analytics_dashboard/courses/views.py (hunk 17 above)
Comment (dsjen) [inline analytics_dashboard/courses/views.py]: Removed.

File: analytics_dashboard/courses/views.py (hunk 18 above)
Comment (dsjen) [inline analytics_dashboard/courses/views.py]: Interesting shortcut!

File: analytics_dashboard/courses/views.py (hunk 14 above)
Comment (dsjen) [inline analytics_dashboard/courses/views.py]: I've refactored the date range calculation into the base presenter class.

File: analytics_dashboard/courses/views.py (hunk 15 above)
Comment (dsjen) [inline analytics_dashboard/courses/views.py]: Default value now set in the base presenter class.

File: analytics_dashboard/courses/presenters.py [hunk 19]
```diff
@@ -4,22 +4,27 @@
...
-    def __init__(self, timeout=5):
+    def __init__(self, course_id, timeout=5):
         # API client
         self.client = Client(base_url=settings.DATA_API_URL,
                              auth_token=settings.DATA_API_AUTH_TOKEN,
                              timeout=timeout)
+        self.course_id = course_id
+        self.course = self.client.courses(self.course_id)
         self.date_format = Client.DATE_FORMAT
+        self.date_time_format = Client.DATE_FORMAT + 'T%f'
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py]: Review/merge https://github.com/edx/edx-analytics-data-api-client/pull/4 and you can use `Client.DATETIME_FORMAT`.

File: analytics_dashboard/courses/presenters.py [hunk 20]
```diff
@@ -122,9 +148,6 @@ def get_geography_data(self):
     def get_summary(self):
         """
         Returns the summary information for enrollments.
-
-        Arguments:
-            course_id (str): ID of the course to retrieve summary information of.
         """
 
         # Establish default return values
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py:154]: Rebase on master. Only `enrollment_change_last_7_days` remains. (Expand the code view. I am referring to the dict declared on line 131/154.)

File: analytics_dashboard/courses/presenters.py [hunk 21]
```diff
@@ -89,18 +118,15 @@ def get_summary(self, course_id):
 class CourseEnrollmentPresenter(BasePresenter):
     """ Presenter for the course enrollment data. """
 
-    def __init__(self, course_id, timeout=5):
-        super(CourseEnrollmentPresenter, self).__init__(timeout)
-        self.course_id = course_id
-
-    def get_trend_data(self, start_date=None, end_date=None):
-        return self.client.courses(self.course_id).enrollment(start_date=start_date, end_date=end_date)
+    def get_trend_data(self, end_date=None, interval=None):
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py]: Consider using `num_days` instead of `interval`. When I think of intervals I think of the amount of time between data points, not the length of the trend.

**Comment by clintonb** [comment]
Rebase and I can take another look. So far, this looks great.

File: analytics_dashboard/courses/presenters.py (hunk 19 above)
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py]: Merged. If you upgrade to API client 0.2.1 you can replace this with `Client.DATETIME_FORMAT`.

File: analytics_dashboard/courses/presenters.py (hunk 21 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py]: Done!

**Comment by dsjen** [comment]
I've rebased against master and updated to version 0.2.1 of the api client.

File: acceptance_tests/test_course_engagement.py [hunk 22]
```diff
@@ -48,3 +51,46 @@ def test_student_activity(self):
...
+        graph_selector = '#engagement-trend-view'
+        EmptyPromise(
+            lambda: 'Loading Trend...' not in self.page.q(css=graph_selector + ' p').text,
+            "Trend finished loading"
+        ).fulfill()
+        self.assertElementHasContent(graph_selector)
+
+    def test_engagement_table(self):
+        self.page.visit()
+        date_time_format = self.api_client.DATE_FORMAT + 'T%f'
```
Comment (clintonb) [inline acceptance_tests/test_course_engagement.py]: Use `self.api_client.DATETIME_FORMAT`.

File: analytics_dashboard/courses/presenters.py [hunk 23]
```diff
@@ -140,10 +163,9 @@ def get_summary(self):
         if recent_enrollment:
             # Get data for a month prior to most-recent data
             last_enrollment_date = self.parse_date(recent_enrollment[0]['date'])
-            month_before = last_enrollment_date - datetime.timedelta(days=31)
```
Comment (clintonb) [inline analytics_dashboard/courses/presenters.py:143]: Something weird is going on here. This shouldn't show up in the diff since these changes are on master. Weird. It _may_ be a non-issue.

Comment (clintonb) [inline analytics_dashboard/courses/presenters.py:143]: It turns out this may be an issue. `acceptance_tests.test_course_enrollment.CourseEnrollmentActivityTests.test_enrollment_summary_and_graph` is failing. The value displayed for last week's enrollment delta is `None` despite the fact that there is data present for the last 8 months.

File: acceptance_tests/test_course_engagement.py (hunk 22 above)
Comment (dsjen) [inline acceptance_tests/test_course_engagement.py]: Done!

File: analytics_dashboard/courses/presenters.py (hunk 23 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py:143]: Good catch!  Fixed now.

**Comment by clintonb** [comment]
:+1:

File: analytics_dashboard/courses/presenters.py (hunk 3 above)
Comment (rocha) [inline analytics_dashboard/courses/presenters.py:95]: Could you add a comment about why it is necessary? Thanks.

File: analytics_dashboard/static/js/engagement-content-main.js [hunk 24]
```diff
@@ -6,5 +6,75 @@
...
+        // colors are chosen to be color-blind accessible
+        var settings =  [
+            {
+                key: 'weekEnding',
+                title: gettext('Week Ending'),
+                type: 'date'
+            },{
+                key: 'any',
+                title: gettext('Active Students'),
+                color: '#8DA0CB',
```
Comment (rocha) [inline analytics_dashboard/static/js/engagement-content-main.js:20]: I probably asked this before... Is is possible to move the color values to the style sheets? I'll love to have all the presentation attributes in one place. Definitely not a blocker.

**Comment by rocha** [comment]
Looks good to me. Can you add the comment I requested before merging? Thanks

:rocket: :moon: :boom:

File: analytics_dashboard/static/js/engagement-content-main.js (hunk 24 above)
Comment (dsjen) [inline analytics_dashboard/static/js/engagement-content-main.js:20]: I really like this idea.  Unfortunately, I don't see a interface to add in the css on the line and points via the nvd3 api.  I think I could probably put something together using selections, etc.  If it's ok, I'll punt on this for now.

File: analytics_dashboard/courses/presenters.py (hunk 3 above)
Comment (dsjen) [inline analytics_dashboard/courses/presenters.py:95]: Ok, done!
